### ISSUE TITLE
Sample Validaton Datasets -- What is Augraphy Trying to Reproduce?

### ISSUE THREAD

**Issue by jboarman** [sparkfish/augraphy#43]
In general, **Augraphy** is trying to simplify the process of creating synthetic realistic datasets using only ground truth documents.

Often, training data is not accompanied by clean ground truth sources, which leads to inaccurate training and severely limited volumes of available training data.  By starting with clean ground truth data, training sets can be created that represent _printed, scanned, copied_ and _faxed_ documents encountered in the real world AND have 100% accurate training data.

In order to recreate data from these real-world scenarios, we need to create a validation set that is inspired by examples from the real world.  Below are sources that may serve as useful source material for attempting to use Augraphy to reproduce the styles and detail seen in these data sets.

# Real-World Data Sets

* **RVL-CDIP** dataset consists of 400,000 B/W low-resolution (~100 DPI) images in 16 classes, with 25,000 images per class
https://www.cs.cmu.edu/~aharley/rvl-cdip/

* **NIST-SFRS (Structured Forms Reference Set)** consists of 5,590 pages of binary, black-and-white images of synthesized documents from 12 different tax forms from the IRS 1040 Package X for the year 1988. These include Forms 1040, 2106, 2441, 4562, and 6251 together with Schedules A, B, C, D, E, F, and SE.
https://www.nist.gov/srd/nist-special-database-2

* **Tobacco3482** dataset from Kaggle offers 10 different classes of forms, letters, reports, etc.
https://www.kaggle.com/patrickaudriaz/tobacco3482jpg

* **FUNSD (Form Understanding Noisy Scanned Documents)** dataset on Kaggle comprises 199 real, fully annotated, scanned forms that are noisy and vary widely in appearance.
https://www.kaggle.com/sharmaharsh/form-understanding-noisy-scanned-documentsfunsd

* **Randomly Collected Documents** is a Google Drive share that contains randomly selected public domain documents.
https://drive.google.com/drive/folders/1JMwmRko1gZ_VYtwXkP7CXPPztsNa_3nv?usp=sharing

# Synthentic Data Sets

* **NoisyOffice** data set from University of California, Irvine contains noisy grayscale printed text images and their corresponding ground truth for both real and simulated documents with 4 types of noise: folded sheets, wrinkled sheets, coffee stains, and footprints.  For each type of font, one type of Noise: 17 files * 4 types of noise = 72 images.
https://archive.ics.uci.edu/ml/datasets/NoisyOffice

* **DDI-100 (Distorted Document Images)** is a synthetic dataset by Ilia Zharikov (ilya.zharikov@phystech.edu) et al based on 7000 real unique document pages and consists of more than 100000 augmented images. Ground truth comprises text and stamp masks, text and characters bounding boxes with relevant annotations.
https://arxiv.org/abs/1912.11658
https://github.com/machine-intelligence-laboratory/DDI-100/tree/master/dataset
https://paperswithcode.com/paper/ddi-100-dataset-for-text-detection-and

**Comment by kwcckw** [comment]
Creating the style should be feasible, take this example:
https://www.kaggle.com/patrickaudriaz/tobacco3482jpg
![image](https://user-images.githubusercontent.com/52521979/127076094-31769a38-5a7b-42db-a27f-f55f0e52c868.png)

From here we can see several matching augmentations, such as page borders, ink bleed, dusty ink and etc. But the content itself could be a problem since it might not be possible if we want to create each of those contents on our own. So maybe we can find some open source digital documents and apply the relevant augmentations to them so that their style is matched? 

For example, apply similar style's agumentations to the letter below :

![image](https://user-images.githubusercontent.com/52521979/127076912-1003656c-7282-411a-a929-f7fe033822aa.png)

**Comment by jboarman** [comment]
> we can see several matching augmentations, such as page borders, ink bleed, dusty ink and etc. But the content itself could be a problem since it might not be possible if we want to create each of those contents on our own

Good point.  We definitely want to have simple and clean inputs that would be representative of something one might have in their ground truth dataset.  In theory, a practitioner would generate the input from text so that the ground truth can be programmatically known in advance, much like you are doing in your "Lorem ipsum dolo" sample notebook.  But, for our test suite, we just need some simple and clean images that could plausibly be generated from a script.

The other consideration is the inclusion of images like the provided sample.

**Comment by kwcckw** [comment]
Okay, so i think we can do it either way. If the source document is having certain format, such as formal email, quotation, newspaper and etc, then we need to get digital copy of those documents, or at least having lines or squares to reproduce the document in similar format. Otherwise, generate text from code would be a better choice since it would be more straightforward.

**Comment by proofconstruction** [comment]
It shouldn't be too hard to write some functions that _produce_ different documents given some strings. We can already easily [write strings into images](https://code-maven.com/create-images-with-python-pil-pillow), and if we had a decent specification of some different kinds of documents (where lines should be drawn, where different areas of text should go), we could turn these into classes that produce ground-truthed documents, which could then be fed into Augraphy to produce the larger training set, which will all have ground-truths.

This collection of document-generation classes should probably be its own package though.

**Comment by jboarman** [comment]
I need your help.

We need to identify 5-10 archetypes, or sample documents, that we aim to reproduce using Augraghy.  For each type of sample document, we need to then create a clean version that closely resembles the original document.  This will then allow us to validate how well Augraphy can reproduce the noise we hope to be producing.

For each sample archetype, we can produce a [notebook that demonstrates using Augraphy](https://colab.research.google.com/drive/1yqrR7QdSd_aP4-R867f_Oxrivg4qlwix?usp=sharing#scrollTo=FWrjUmY9JEGl) to go full circle, showing how that noise can be synthesized.  The initial notebooks might not look perfect, but they could serve as a sort of starting point for a competition that betters Augraphy towards faithful reproductions of each of these archetypes.

**HELP:** Can you help identify example documents that show variety in the types of noise and issue seen in these documents?

(Don't worry about creating the clean reproduction -- let's sort out the archetypes first!)

# Original Target Document
This example document comes from the RVL-CDIP dataset and represents the kind of noisy document that we intend to reproduce with Augraphy:

![image](https://user-images.githubusercontent.com/1108065/130364104-251fdafa-5928-441b-be8d-2765fbe0c3c7.png)

# Clean Reproduction
This is a [reproduction made using Google Documents](https://docs.google.com/document/d/1JAtwUiyHRktbzNf0l9wceysVauDF80xyoGFvF_KGyM0/edit) of what the original document might look like prior to being deteriorated by copying and scanning, etc:

![image](https://user-images.githubusercontent.com/1108065/130365478-00cfe27c-2550-4923-abbc-1d04a8889ede.png)

**Comment by kwcckw** [comment]
So right now we are only looking at RVL-CDIP dataset right? I need some time to download it since the dataset size is large. As a first step,  I will check through those images and do a compilation of augmentations and their sample image in the dataset. Once i getting sufficient number of samples, i will share the details again here.

**Comment by jboarman** [comment]
> So right now we are only looking at RVL-CDIP dataset right?

I’d say any of the sources listed in this issue could be used as a reference for identifying sample documents.  It will likely take a lot of process of elimination work to narrow it down.

**Comment by proofconstruction** [comment]
I started a [collection of representative images](https://drive.google.com/drive/folders/140CY0f9RmZE88z722s669QhPqQOgAZJK) we might reproduce. The five currently there were pulled from the [Randomly Collected Documents](https://drive.google.com/drive/folders/1JMwmRko1gZ_VYtwXkP7CXPPztsNa_3nv?usp=sharing) set.

They're numbered 1.jpg, 2.jpg, and so on in the folder, and the notes below correspond to those file names.

### Images
Here's why I like them, as a list of features we can aim to reproduce:

#### Image 1
1. The text is faded in many places and to varying degrees
2. the source document was slightly rotated when scanning
3. there is a page border effect on one side
4. some hand-written text is visible on the bottom, and faded on the side.

#### Image 2
1. distortion in the top-border
2. hand-written redaction done with a marker (new effect?)
3. visible print lines
4. "snow"/random noise in places around the image

#### Image 3
1. the bleedthrough effect is very pronounced here, and importantly *the reverse side does not contain the same text as the front*
2. distorted page edges
3. lighting gradient on page edges
4. stains
5. noise along the bottom edge
6. the texture of the page is apparent

#### Image 4
1. tons of noise
2. redactions
3. complex page border geometry
4. multiple styles of handwritten text
5. stamps!
 
#### Image 5
1. page borders
2. redactions
3. punched holes on the top edge!
4. noise along the top and bottom, noise in the top right
5. more handwritten text, some of which overlaps printed text

### Replication
I think most of what I've listed here is reproducible with existing augmentations; in particular, I think all of the handwriting can be produced by (some straightforward generalization of) the new pencil augmentation, and we can already do serviceable page borders, fading, noise, brightness texturization, bleedthrough, and so on wherever we like.

What we don't have though, is a way to mimic stamps, redaction, or binder holes. We should be able to do binder holes synthetically rather easily, and the challenge with redaction would be correctly placing the mark on the text, but I'm not sure about how to approach stamps without having a collection of them included like the paper textures are. Maybe stamps are out of scope.

**Comment by kwcckw** [comment]
Here's the compiled list of images by looking only at rvl-cdip dataset `imagesA` folder :
https://drive.google.com/drive/folders/17Z4VRqc1w9ZKpjoO3dulSqXTpxPVutMN?usp=sharing

Some are not fully reproducible, i included them here since they look interesting enough.


`img_67.png` (Not fully reproducible, some augmentations not exist)
https://drive.google.com/file/d/1xJFoZ3XTA45MZ6LtC7sWMWOUGeUhHSKR/view?usp=sharing
1. Badphotocopy
2. Dirtydrum (half page with faded effect)
3. Page border (unique in this case because it is not in the page border)
4. Scribbles (Numbers) (we don't have this effect yet)
5. Punch holes (we don't have this effect yet)
6. Clips' marks (we don't have this effect yet)


`img_299.png` (Fully reproducible with some changes in the current code)
https://drive.google.com/file/d/17ev5mpuqMWd3l90nPCH9wbkNyekAhzB7/view?usp=sharing
1. Badphotocopy (Very random noises)
2. Page border on top


`img_34123.png` (Fully reproducible with some changes in the current code)
https://drive.google.com/file/d/1T34aLs58_V6xaHB0WfpttFMyyXYDQFB5/view?usp=sharing
1. Badphotocopy (surrounding the page at page border only)


`img_34212.png`  (Fully reproducible)
https://drive.google.com/file/d/1ctNWE9Fx7zOlBNfKoFCnRRhDuABu-OAL/view?usp=sharing
1. Dirty roller
2. Page borders


`img_34462.png` (Not fully reproducible, some augmentations not exist)
https://drive.google.com/file/d/1IKAlc0ejGLFKiuKMhnSSPbOfXtM95hEU/view?usp=sharing
1. Scribbles (words)(we don't have this effect yet)
2. Page borders


`img_109076.png` (Not fully reproducible, some augmentations not exist)
https://drive.google.com/file/d/1u1FSbsNYAXcfp3GLD_WkjmSuJpbSt6aB/view?usp=sharing
1. Incomplete page (torn off?) (we don't have this effect yet)
2. Page border
3. Bad photocopy
4. Punch holes (we don't have this effect yet)


`img_136566.png` (Fully reproducible)
https://drive.google.com/file/d/1MbcKbPDUNQlcVwFw4GoIPCZLJHcmUXdt/view?usp=sharing
1. Book binding 
2. Dirty oller (minor)


`img_166297.png` (Fully reproducible)
https://drive.google.com/file/d/1Z5RAP_DkUMnVAqd2R0g4m2tEGHvjEIGK/view?usp=sharing
1. Letterpress
2. Page border (very minor at bottom)


`img_171248.png` (Not fully reproducible, some augmentations not exist)
https://drive.google.com/file/d/1iYzC3NbfIMMx7GcJ6JQUgOJouC_NM1ek/view?usp=sharing
1. Blobs of noises (we don't have this effect yet)


`img_179298.png` (Fully reproducible)
https://drive.google.com/file/d/1kRQarbuQF96tVUndHQUAyBqSFfW3qs0T/view?usp=sharing
1. Page border
2. Bookbinding

**Comment by shaheryar1** [comment]
By Looking at **Resume** category of _Tobacco3482_ dataset, I analyzed that almost all of the resumes are grayscale images with plain text on them therefore the `Ink phase` is very useful for reproducing such documents.

Here is a small [list ](https://drive.google.com/drive/folders/1GTVraYhwwYdryGFGv65Yb1uDrymnt04_?usp=sharing) of some resumes extracted from dataset which collectively represents the noise in overall dataset.



https://drive.google.com/file/d/1OXdjEhbNE6moWaia0ORxUGPeU9tx69J8/view?usp=sharing
1. InkBleed (with high intensity ranges)
2. LowInkLine (Periodic)
3. DustyInk
4. PageBorder (Left)


https://drive.google.com/file/d/108OO9dGNF3FZmv-PpMFrrDYUfDBGZzD9/view?usp=sharing
1. InkBleed (low intensity)
2. Punch Holes ( not present in augraphy, discussion continues in #62 )
3. Page Border (top and bottom)


https://drive.google.com/file/d/1AdZyOpMtecIPRswVd6kwDooToKjY8G3u/view?usp=sharing
1. InkBleed (with high intensity ranges)
2. Jpeg Compression
3. Subtle Noise
4. PencilScribbles (of very small size and full-black color)
5. PageBorder on top (with a margin and diminish effect)
6. Applying noisy-blobs in a straight vertical line of the page (similar to Folding effect)
